src/tests/btree_helpers/shadow_map.hpp

```diff
@@ -7,36 +7,48 @@ template < typename K, typename V >
 class ShadowMap {
 private:
     std::map< K, V > m_map;
+    RangeScheduler m_range_scheduler;
+    using mutex = iomgr::FiberManagerLib::shared_mutex;
+    mutex m_mutex;
 
 public:
+    ShadowMap(uint32_t num_keys) : m_range_scheduler(num_keys) {}
+
     void put_and_check(const K& key, const V& val, const V& old_val, bool expected_success) {
+        std::lock_guard lock{m_mutex};
         auto const [it, happened] = m_map.insert(std::make_pair(key, val));
         ASSERT_EQ(happened, expected_success) << "Testcase issue, expected inserted slots to be in shadow map";
         if (!happened) {
             ASSERT_EQ(old_val, it->second) << "Put: Existing value doesn't return correct data for key: " << it->first;
         }
+        m_range_scheduler.put_key(key.key());
     }
 
     void range_upsert(uint64_t start_k, uint32_t count, const V& val) {
+        std::lock_guard lock{m_mutex};
         for (uint32_t i{0}; i < count; ++i) {
             K key{start_k + i};
             V range_value{val};
             if constexpr (std::is_same_v< V, TestIntervalValue >) { range_value.shift(i); }
             m_map.insert_or_assign(key, range_value);
         }
+        m_range_scheduler.put_keys(start_k, start_k + count - 1);
     }
 
     void range_update(const K& start_key, uint32_t count, const V& new_val) {
+        std::lock_guard lock{m_mutex};
         auto const start_it = m_map.lower_bound(start_key);
         auto it = start_it;
         uint32_t c = 0;
         while ((it != m_map.end()) && (++c <= count)) {
             it->second = new_val;
             ++it;
         }
+        m_range_scheduler.remove_keys_from_working(start_key.key(), start_key.key() + count - 1);
     }
 
     std::pair< K, K > pick_existing_range(const K& start_key, uint32_t max_count) const {
+        std::lock_guard lock{m_mutex};
```

Comment:
wondering if we can use std::shared_mutex? Also I believe the test will run in fiber, so would prefer using iomanager mutex and shared_mutex

Reply:
Ack